workspaces/src/rpc/query.rs

```diff
@@ -0,0 +1,361 @@
+                account_id: self.account_id,
+                prefix: StoreKey::from(self.prefix.map(Vec::from).unwrap_or_default()),
+                include_proof: false,
+            },
+        })
+    }
+
+    fn from_response(resp: <Self::Method as RpcMethod>::Response) -> Result<Self::Output> {
+        match resp.kind {
+            QueryResponseKind::ViewState(state) => Ok(tool::into_state_map(state.values)),
+            _ => Err(RpcErrorCode::QueryReturnedInvalidData.message("while querying state")),
+        }
+    }
+}
+
+impl<'a> Query<'a, ViewState> {
+    pub(crate) fn view_state(client: &'a Client, id: &AccountId) -> Self {
+        Self::new(
+            client,
+            ViewState {
+                account_id: id.clone(),
+                prefix: None,
+            },
+        )
+    }
+
+    /// Set the prefix for viewing the state.
+    pub fn prefix(mut self, value: &[u8]) -> Self {
+        self.method.prefix = Some(value.into());
+        self
+    }
+}
+
+impl ProcessQuery for ViewAccessKey {
+    type Method = methods::query::RpcQueryRequest;
+    type Output = AccessKey;
+
+    fn into_request(self, block_reference: BlockReference) -> Result<Self::Method> {
+        Ok(Self::Method {
+            block_reference,
+            request: QueryRequest::ViewAccessKey {
+                account_id: self.account_id,
+                public_key: self.public_key.into(),
+            },
+        })
+    }
+
+    fn from_response(resp: <Self::Method as RpcMethod>::Response) -> Result<Self::Output> {
+        match resp.kind {
+            QueryResponseKind::AccessKey(key) => Ok(key.into()),
+            _ => Err(RpcErrorCode::QueryReturnedInvalidData.message("while querying state")),
+        }
+    }
+}
+
+impl ProcessQuery for ViewAccessKeyList {
+    type Method = methods::query::RpcQueryRequest;
+    type Output = Vec<AccessKeyInfo>;
+
+    fn into_request(self, block_reference: BlockReference) -> Result<Self::Method> {
+        Ok(Self::Method {
+            block_reference,
+            request: QueryRequest::ViewAccessKeyList {
+                account_id: self.account_id,
+            },
+        })
+    }
+
+    fn from_response(resp: <Self::Method as RpcMethod>::Response) -> Result<Self::Output> {
+        match resp.kind {
+            QueryResponseKind::AccessKeyList(keylist) => {
+                Ok(keylist.keys.into_iter().map(Into::into).collect())
+            }
+            _ => Err(RpcErrorCode::QueryReturnedInvalidData.message("while querying state")),
+        }
+    }
+}
+
+impl ProcessQuery for GasPrice {
+    type Method = methods::gas_price::RpcGasPriceRequest;
+    type Output = Balance;
+
+    fn into_request(self, block_ref: BlockReference) -> Result<Self::Method> {
+        let block_id = match block_ref {
+            BlockReference::BlockId(block_id) => Some(block_id),
+            BlockReference::Finality(_finality) => None,
+            _ => panic!(""),
```

Comment:
Why is this necessary that can't be overcome with a local type? If it's necessary, can you at least add a message?

Reply:
oops, that panic should've returned an error. Not sure what you mean by local type here. Something like `GasPrice { block_id: ... }`? All this is doing is reusing the common `block_ref` in every `Query` for the block_id portion. The finality is default case when no `block_hash` or `block_height` function is used, but definitely should leave a comment saying why this is the case here 👍

Reply:
I just mean, there is no way in the code to construct this, so we could just restrict by having an enum that doesn't contain that variant to avoid it completely by the type system rather than returning an error?

Reply:
Ahh, I see what you mean. I did something like that initially, with another associated type `Input` in the `ProcessQuery` trait, but it just complicated things, when `GasPrice` was the only extra case. Like, we'd have to add in a lot more boilerplate code to be specific about just having either a `BlockReference` vs just a `BlockId` for gas price, so I opted not to do that for now until we have at least 2 or 3 that depended on that behavior. It's all internal anyways, so wouldn't be exposed if we decide to change it later